### PR TITLE
Fix CAST(JSON as ROW(ARRAY))

### PR DESCRIPTION
**Pull request by mbasmanova**
Summary:
CAST(JSON as ROW(ARRAY()) used to fail with

```
OUT_OF_ORDER_ITERATION: Objects and arrays can only be iterated when they are first encountered.
```

According to simdjson documentation, https://github.com/simdjson/simdjson/blob/master/doc/basics.md, it is not allowed to store object values for later processing. These must be consumed or copied before proceeding.

Also, fixed behavior when JSON object contains duplicate keys. Presto throws, but previous implementation used to allow duplicates.

Also, fix the test to actually verify JSON objects with mixed case keys.

Differential Revision: D56013293

### CODE REVIEW

File: velox/functions/prestosql/types/JsonType.cpp [hunk 1]
```diff
@@ -856,30 +856,46 @@ struct CastFromJsonTypedImpl {
...
+        // Mapping from lower-case field names of the target RowType to their
+        // indices.
+        folly::F14FastMap<std::string, column_index_t> fieldIndices;
+        for (auto i = 0; i < rowType.size(); ++i) {
+          auto key = rowType.nameOf(i);
+          boost::algorithm::to_lower(key);
+          fieldIndices[key] = i;
+        }
+
+        std::vector<bool> foundIndices(rowType.size(), false);
```
Comment (Yuhta) [inline velox/functions/prestosql/types/JsonType.cpp:870]: We could initialize `fieldIndices` value to -1 then we don't need this second memory.

**Review by Yuhta — APPROVED** [review]
LGTM

File: velox/functions/prestosql/types/JsonType.cpp (hunk 1 above)
Comment (mbasmanova) [reply velox/functions/prestosql/types/JsonType.cpp]: Sounds good. Updated.

**Comment by mbasmanova** [comment]
@Yuhta Jimmy, thank you for review. Updated to address your comments.

**Review by Yuhta — APPROVED** [review]

**Review by xiaoxmeng — APPROVED** [review]
@mbasmanova thanks!

**Comment by kgpai** [comment]
Please rebase to main after https://github.com/facebookincubator/velox/pull/9451 lands.

**Pull request closed without merging**